util.c

```diff
@@ -1932,6 +1932,55 @@ int mdmon_running(const char *devnm)
 	return 0;
 }
 
+/* wait_for_mdmon_control_socket() - Waits for mdmon control socket
+ * to be created within specified time.
+ * @devnm: Device for which mdmon control socket should start.
+ *
+ * In foreground mode, when mdadm is trying to connect to control
+ * socket it is possible that the mdmon has not created it yet.
+ * Give some time to mdmon to create socket. Timeout set to 10 sec.
+ */
+int wait_for_mdmon_control_socket(char *devname)
```

Comment:
I see that the code in this function is a copy of the code from `connect_monitor`
In this form it cannot be accepted, we should not repeat same code. It must be optimized some way.

You can extract similar parts to helper functions, for example (not tested):
```c
mdadm_status_t get_mdmon_socket_file(char *devnm, char *buf)
{
        char *c;

	if (is_subarray(devname)) {
		devname++;
		c = strchr(devname, '/');
		if (!c)
			return MDADM_STATUS_ERROR;
               c++;
	} else
		c = devname;
	snprintf(buf, PATH_MAX, "%s/%s.sock", MDMON_DIR, c);
}
```

This is just a part of function, other places should be optimized too.

Reply:
I will try to optimize the similar code.

Reply:
@mtkaczyk I have pushed in the new changes. Please share your feedback.

FYI, while testing these changes, I noticed one thing that we might not need subarray check in case of container or specficially in case of wait_for_mdmon_control_socket() function context, because container name itself wont be in the subarray name format. Container name is in the format of "mdX" or "mdY" while subarray  can be of the format of "/md20/X" or "/md20/Y". 
I am still new to these terminologies or naming conventions, so I am not sure if the conclusion I have come to is correct or not.

Reply:
Thank you :)

Reply:
> FYI, while testing these changes, I noticed one thing that we might not need subarray check in case of container or specficially in case of wait_for_mdmon_control_socket() function context, because container name itself wont be in the subarray name format. Container name is in the format of "mdX" or "mdY" while subarray can be of the format of "/md20/X" or "/md20/Y". I am still new to these terminologies or naming conventions, so I am not sure if the conclusion I have come to is correct or not.

that is true, this is simplification for case when you are opening subarray only. This can be read from /sys/block/<subarray>/md/version. if you don't need this logic now fell free to remove. No need to keep dead code.